engine/language-server/src/main/scala/org/enso/languageserver/boot/ComponentSupervisor.scala

```diff
@@ -0,0 +1,20 @@
+package org.enso.languageserver.boot
+
+/** A wrapper around an existing component/resource that can be added with a delay and ensured
+  * to be closed safely.
+  */
+class ComponentSupervisor extends AutoCloseable {
+  private var component: AutoCloseable = null
+
+  def registerService(component: AutoCloseable): Unit = {
+    assert(this.component == null, "can't register component twice")
```

Comment:
AFAIK it works because the single component initialization is guaranteed by the `LockedInitialization` implementation.

Reply:
Yeah, this is just a precaution in case someone wants to use in a wrong way.